src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/CreateDream.kt

```diff
@@ -28,12 +37,29 @@ class CreateDream(
       .flatMap { dream ->
         findUserById.execute(userId)
           .toEither { DreamCreationErrors.CreatorNotFound(userId) }
-          .map { user ->
-            dreamRequest.tags.forEach { tag ->
-              createTag.execute(tag).map { dream.tags.add(it) }
-            }
-            dao.save(dream.toEntity(user))
+          .flatMap { user ->
+            dreamRequest.tags
+              .map(createTag::execute).k()
+              .sequence(Either.applicative()).fix()
+              .map { tags: ListK<Tag> ->
+                val dreamWithTags = dream.copy(tags = tags)
+                dao.save(dreamWithTags.toEntity(user))
+              }
           }
       }
       .map { it.toDomain() }
+
+  fun alternativeVersion(
+    dreamRequest: DreamRequest,
+    userId: Long,
+    currentTime: Long = System.currentTimeMillis()
+  ): Either<DreamCreationErrors, Dream> = Either.monad<DreamCreationErrors>().binding {
+    val dream = dreamRequest.toDomain(userId)
+    val validatedDream = validateDream(dream, currentTime).bind()
+    val user = findUserById.execute(userId).toEither { DreamCreationErrors.CreatorNotFound(userId) }.bind()
+    val tags = dreamRequest.tags.map(createTag::execute).k().sequence(Either.applicative()).fix().bind()
+    val dreamWithTags = validatedDream.copy(tags = tags)
+    val dreamEntity = dreamWithTags.toEntity(user)
+    dao.save(dreamEntity).toDomain()
+  }.fix()
```

Comment:
This is another version of the method above. This time we are using monad comprehensions to avoid heavy indentation. You can read more about this functional pattern [here](https://arrow-kt.io/docs/patterns/monad_comprehensions/)

Reply:
I think that I prefer the version above :thinking:

Reply:
Why do you prefer the first version? Is because of the `bind()` method?

Reply:
Nop, I saw the `arrow` documentation and the `bind()` is not the problem haha But for me I see it more clear with the other version, but as I said before there could be cases that I would prefer this version.